share/p2p/shrex/client.go

```diff
@@ -0,0 +1,172 @@
+package shrex
+
+import (
+	"errors"
+	"fmt"
+
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"golang.org/x/net/context"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/ipld"
+	share_p2p_v1 "github.com/celestiaorg/celestia-node/share/p2p/v1/pb"
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+var (
+	errInvalidStatusCode = errors.New("INVALID")
+	errNotFound          = errors.New("NOT_FOUND")
+	errServerInternal    = errors.New("SERVER_INTERNAL")
+	errConnRefused       = errors.New("CONN_REFUSED")
+	errNotAvailable      = errors.New("none of the peers has requested data")
+)
+
+// Client implements Getter interface, requesting data from remote peers
+type Client struct {
+	host           host.Host
+	defaultTimeout time.Duration
+
+	// fake peers for testing purposes
+	testPeers []peer.ID
+}
+
+// NewClient creates a new shrEx client
+func NewClient(host host.Host, defaultTimeout time.Duration) *Client {
+	return &Client{
+		host:           host,
+		defaultTimeout: defaultTimeout,
+	}
+}
+
+// GetBlobByNamespace request shares with option to collect proofs from remote peers using shrex protocol
+func (c *Client) GetBlobByNamespace(ctx context.Context, root *share.Root, nID namespace.ID) (*share.Blob, error) {
+	peers := c.testPeers
+	if len(peers) == 0 { //nolint:staticcheck
+		//TODO: collect peers from discovery here
+	}
+
+	for _, peer := range peers {
```

Comment:
I THINK it should be retried across all peers until all of them are excluded or the data can be returned, like in shrex/eds, but this isn't clear in the ADR. @Wondertan can you clarify?